Live activities push to start (#6930)
braze-inc/braze-docs: _docs/_developer_guide/platform_integration_guides/swift/live_activities/live_activities.md

{% sdk_min_versions swift:5.11.0 %}

Additional prerequisites include:

- Live Activities are only available for iPhones and iPads on iOS 16.1 and later. To use this feature, ensure that your project is targeting this iOS version.
- The `Push Notification` entitlement must be added under **Signing & Capabilities** in your Xcode project.

> - Remotely starting a Live Activity via push notifications was introduced in version 8.Y.Z of the Braze SDK and requires iOS 17.2 and later.

Author: - [ ] Update version
```suggestion
- Remotely starting a Live Activity via push notifications was introduced in version 8.1.0 of the Braze SDK and requires iOS 17.2 and later.
```

1. [Create the Live Activity.](#developing) Develop the Live Activity UI using WidgetKit and SwiftUI. Initialize a Live Activity object with the relevant data models for your static and dynamic states.<br><br>

2. Start the Live Activity.<br><br>

>    a. [via push-to-start](#starting) Register a Live Activity type with the Braze SDK using the [`registerPushToStart`](<https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/liveactivities-swift.class/launchactivity(pushtokentag:activity:fileid:line:)>) method and remotely start an activity of that type using our [`/messages/live_activity/start` endpoint]({{site.baseurl}}/api/endpoints/messaging/live_activity/start).<br><br>

Author: - [ ] Update this link after releasing the updated docs:
```suggestion
   a. [via push-to-start](#starting) Register a Live Activity type with the Braze SDK using the [`registerPushToStart`](<https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/liveactivities-swift.class/update-method>) method and remotely start an activity of that type using our [`/messages/live_activity/start` endpoint]({{site.baseurl}}/api/endpoints/messaging/live_activity/start).<br><br>
```

}
```

> ### Step 2: Register the activity

Reviewer: ```suggestion
### Step 2: Start the activity
```

@jerielng would it make sense to call this start, or is register the proper terminology here.

Author: Yep I feel like `Start` would be better, although we'd need to make a couple of changes to the content. I'll take a dive into that.

First, choose how you want to register your activity:

> - **Remote:** Use the [`registerPushToStart`](http://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/liveactivities-swift.class/registerpushtostart(fortype:name:)) method, then start an activity using the [`/messages/live_activity/start`]({{site.baseurl}}/api/endpoints/messaging/live_activity/start) endpoint.

Reviewer: ```suggestion
- **Remotely:** Use the [`registerPushToStart`](http://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/liveactivities-swift.class/registerpushtostart(fortype:name:)) method, then start an activity using the [`/messages/live_activity/start`]({{site.baseurl}}/api/endpoints/messaging/live_activity/start) endpoint.
```

> - **Local:** Use the [`launchActivity`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/liveactivities-swift.class/launchactivity(pushtokentag:activity:fileid:line:)) method to create push tokens for the Braze SDK to manage.

Reviewer: ```suggestion
- **Locally:** Use the [`launchActivity`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/liveactivities-swift.class/launchactivity(pushtokentag:activity:fileid:line:)) method to create push tokens for the Braze SDK to manage.
```

just a suggestion (here and tabs below)

{% endalert %}

###### Example

In the following example, the `LiveActivityManager` class handles Live Activity objects. Then, the `registerPushToStart` method registers `SportActivityAttributes`:

```swift
import BrazeKit

#if canImport(ActivityKit)
  import ActivityKit
#endif

class LiveActivityManager {

  @available(iOS 17.2, *)
  func registerActivityType() {
    // This method returns a Swift background task.
    // You may keep a reference to this task if you need to cancel it wherever appropriate, or ignore the return value if you wish.
    let pushToStartObserver: Task = Self.braze?.liveActivities.registerPushToStart(
      forType: Activity<SportsActivityAttributes>.self,
      name: "SportsActivityAttributes"
    )
  }

}
```
{% endtab %}

{% tab local %}
You can use [Apple's ActivityKit framework](https://developer.apple.com/documentation/activitykit) to get a push token, which the Braze SDK can manage for you. This allows you to update Live Activities through the Braze API, as Braze will send the push token to the Apple Push Notification service (APNs) on the backend.

1. Create an instance of your Live Activity implementation using Apple’s ActivityKit APIs.

> 2. Set the `pushType` parameter as `.token`. 

Reviewer 2: Good catch! Yes, for some reason there can be issues when trying to applying markdown styling like `**` inside a link + using reference-styling linking. So changing it to this instead, fixed the issue: `[text](url)`.

}
```

## Step 2a: Registering and Starting a Live Activity remotely {#starting}

{% alert note %}

> Live Activity push-to-start features require iOS 17.2+ and version 8.Y.Z of the Braze Swift SDK.

Reviewer: todo: update after version is finalized

Once you have registered the Live Activity, the Braze SDK will extract and observe changes in the push tokens.

#### Example

> For our example, we’ll create class called `LiveActivityManager` as an interface for our Live Activity objects. Then, we'll set the `pushTokenTag` to `"live-activity-1"`.

Reviewer: ```suggestion
For our example, we’ll create class called `LiveActivityManager` as an interface for our Live Activity objects. Then, we'll set the `pushTokenTag` to `"sports-game-2024-03-15"`.
```

would be good to have a clear example activity ID like this

  
  @available(iOS 16.2, *)
  func createActivity() {
    let activityAttributes = SportsActivityAttributes(gameName: "Superb Owl", gameNumber: "Game 1")
    let contentState = SportsActivityAttributes.ContentState(teamOneScore: "0", teamTwoScore: "0")
    let activityContent = ActivityContent(state: contentState, staleDate: nil)
    if let activity = try? Activity.request(attributes: activityAttributes,
                                            content: activityContent,
      // Setting your pushType as .token allows the Activity to generate push tokens for the server to watch.
                                            pushType: .token) {
      // Register your Live Activity with Braze using the pushTokenTag.
      // This method returns a Swift background task.
      // You may keep a reference to this task if you need to cancel it wherever appropriate, or ignore the return value if you wish.

>       let liveActivityObserver: Task = AppDelegate.braze?.liveActivities.launchActivity(pushTokenTag: "live-activity-1",

Author: ```suggestion
      let liveActivityObserver: Task = AppDelegate.braze?.liveActivities.launchActivity(pushTokenTag: "sports-game-2024-03-15",
```
